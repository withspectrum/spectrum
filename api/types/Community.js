// @flow
const Community = /* GraphQL */ `
	type CommunityChannelsConnection {
		pageInfo: PageInfo!
		edges: [CommunityChannelEdge!]
	}

	type CommunityChannelEdge {
		node: Channel!
	}

	type CommunityMembersConnection @deprecated(reason:"Use the new Community.members type") {
		pageInfo: PageInfo!
		edges: [CommunityMemberEdge!]
	}

	type CommunityMemberEdge @deprecated(reason:"Use the new Community.members type") {
		cursor: String!
		node: User!
	}

	type CommunityMembers {
		pageInfo: PageInfo!
		edges: [CommunityMembersEdge!]
	}

	type CommunityMembersEdge {
		cursor: String!
		node: CommunityMember!
	}

	type CommunityThreadsConnection {
		pageInfo: PageInfo!
		edges: [CommunityThreadEdge!]
	}

	type CommunityThreadEdge {
		cursor: String!
		node: Thread!
	}

	type CommunityMetaData {
		members: Int
		channels: Int
	}

	type SlackImport @deprecated(reason: "Use the slack settings field instead") {
		members: String
		teamName: String
		sent: Date
	}

	type TopAndNewThreads {
		topThreads: [Thread]
		newThreads: [Thread]
	}

	type StripeCard {
		brand: String
		exp_month: Int
		exp_year: Int
		last4: String
	}

	type StripeSource {
		id: ID
		card: StripeCard
		isDefault: Boolean
	}

	type StripeItem {
		id: ID
		amount: Int
		quantity: Int
		planId: String
		planName: String
	}

	type StripeSubscriptionItem {
		created: Int
		planId: String
		planName: String
		amount: Int
		quantity: Int
		id: String
	}

	type StripeDiscount {
		amount_off: Int
		percent_off: Int
		id: String
	}

	type StripeSubscription {
		id: ID
		created: Int
		discount: StripeDiscount
		billing_cycle_anchor: Int
		current_period_end: Int
		canceled_at: Int
		items: [StripeSubscriptionItem]
		status: String
	}

	type StripeInvoice {
		id: ID
		date: Int
		items: [StripeItem]
		total: Int
	}

	type CommunityBillingSettings {
		pendingAdministratorEmail: LowercaseString
		administratorEmail: LowercaseString
		sources: [StripeSource]
		invoices: [StripeInvoice]
		subscriptions: [StripeSubscription]
	}

	type Features {
		analytics: Boolean
		prioritySupport: Boolean
	}
	
	type BrandedLogin {
		isEnabled: Boolean
		message: String
	}

	type Community {
		id: ID!
		createdAt: Date!
		name: String!
		slug: LowercaseString!
		description: String!
		website: String
		profilePhoto: String
		coverPhoto: String
		reputation: Int
		pinnedThreadId: String
		pinnedThread: Thread
    communityPermissions: CommunityPermissions @cost(complexity: 1)
    channelConnection: CommunityChannelsConnection! @cost(complexity: 1)
    members(first: Int = 10, after: String, filter: MembersFilter): CommunityMembers! @cost(complexity: 5, multiplier: "first")
    threadConnection(first: Int = 10, after: String): CommunityThreadsConnection! @cost(complexity: 2, multiplier: "first")
    metaData: CommunityMetaData @cost(complexity: 10)
    invoices: [Invoice] @cost(complexity: 1)
		recurringPayments: [RecurringPayment]
    isPro: Boolean @cost(complexity: 1)
    memberGrowth: GrowthData @cost(complexity: 10)
    conversationGrowth: GrowthData @cost(complexity: 3)
    topMembers: [CommunityMember] @cost(complexity: 10)
    topAndNewThreads: TopAndNewThreads @cost(complexity: 4)
		watercooler: Thread
		brandedLogin: BrandedLogin
		slackSettings: CommunitySlackSettings @cost(complexity: 2)

		hasFeatures: Features
		hasChargeableSource: Boolean
		billingSettings: CommunityBillingSettings

		slackImport: SlackImport @cost(complexity: 2) @deprecated(reason: "Use slack settings field instead")
		memberConnection(first: Int = 10, after: String, filter: MemberConnectionFilter): CommunityMembersConnection! @deprecated(reason:"Use the new Community.members type")
		contextPermissions: ContextPermissions @deprecated(reason:"Use the new CommunityMember type to get permissions")
	}

	extend type Query {
		community(id: ID, slug: LowercaseString): Community
		communities(slugs: [LowercaseString], ids: [ID], curatedContentType: String): [Community]
		communityMember(userId: String, communityId: String): CommunityMember
    topCommunities(amount: Int = 20): [Community!] @cost(complexity: 4, multiplier: "amount")
		recentCommunities: [Community!]

		searchCommunities(string: String, amount: Int = 20): [Community] @deprecated(reason:"Use the new Search query endpoint")
		searchCommunityThreads(communityId: ID!, searchString: String): [Thread] @deprecated(reason:"Use the new Search query endpoint")
	}

	input MembersFilter {
		isOwner: Boolean
		isMember: Boolean
		isBlocked: Boolean
		isPending: Boolean
		isModerator: Boolean
	}

	input MemberConnectionFilter @deprecated(reason: "Use the new MembersFilter input type") {
		isOwner: Boolean
		isMember: Boolean
		isBlocked: Boolean
		isPending: Boolean
		isModerator: Boolean
	}

	input CreateCommunityInput {
		name: String!
		slug: LowercaseString!
		description: String!
		website: String
		file: Upload
		coverFile: Upload
	}

	input EditCommunityInput {
		name: String
		description: String
		website: String
		file: Upload
		coverFile: Upload
		communityId: ID!
	}

	input UpgradeCommunityInput {
		plan: String!
		token: String!
		communityId: String!
	}

	input DowngradeCommunityInput {
		id: String!
	}

	input UpdateAdministratorEmailInput {
		id: ID!
		email: LowercaseString!
	}

	input AddPaymentSourceInput {
		sourceId: String!
		communityId: ID!
	}

	input RemovePaymentSourceInput {
		sourceId: String!
		communityId: ID!
	}

	input MakePaymentSourceDefaultInput {
		sourceId: String!
		communityId: ID!
	}

	input CancelSubscriptionInput {
		communityId: ID!
	}

	input EnableCommunityAnalyticsInput {
		communityId: ID!
	}

	input DisableCommunityAnalyticsInput {
		communityId: ID!
	}
	
	input EnableBrandedLoginInput {
		id: String!
	}

	input DisableBrandedLoginInput {
		id: String!
	}

	input SaveBrandedLoginSettingsInput {
		id: String!
		message: String
	}

	input ImportSlackMembersInput @deprecated(reason: "Slack imports are no longer used, invites sent directly with sendSlackInvites") {
    id: String!
  }

	input SendSlackInvitesInput {
		id: ID!
		customMessage: String
	}

	extend type Mutation {
		createCommunity(input: CreateCommunityInput!): Community
		editCommunity(input: EditCommunityInput!): Community
		deleteCommunity(communityId: ID!): Boolean
		toggleCommunityMembership(communityId: ID!): Community @deprecated(reason:"Use the new addCommunityMember or removeCommunityMember mutations")
		sendSlackInvites(input: SendSlackInvitesInput!): Community
		importSlackMembers(input: ImportSlackMembersInput!): Boolean @deprecated(reason:"Importing slack members is deprecated")
		sendEmailInvites(input: EmailInvitesInput!): Boolean
		pinThread(threadId: ID!, communityId: ID!, value: String): Community
		upgradeCommunity(input: UpgradeCommunityInput!): Community @deprecated(reason:"Use feature level upgrade mutations like enableCommunityAnalytics")
		downgradeCommunity(input: DowngradeCommunityInput!): Community @deprecated(reason:"Use feature level downgrade mutations like disableCommunityAnalytics")
		updateAdministratorEmail(input: UpdateAdministratorEmailInput!): Community
		addPaymentSource(input: AddPaymentSourceInput!): Community
		removePaymentSource(input: RemovePaymentSourceInput!): Community
		makePaymentSourceDefault(input: MakePaymentSourceDefaultInput!): Community
		cancelSubscription(input: CancelSubscriptionInput!): Community
		enableCommunityAnalytics(input: EnableCommunityAnalyticsInput!): Community
		disableCommunityAnalytics(input: DisableCommunityAnalyticsInput!): Community
		enableBrandedLogin(input: EnableBrandedLoginInput!): Community
		disableBrandedLogin(input: DisableBrandedLoginInput!): Community
		saveBrandedLoginSettings(input: SaveBrandedLoginSettingsInput!): Community
	}
`;

module.exports = Community;
