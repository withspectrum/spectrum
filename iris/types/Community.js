const Community = /* GraphQL */ `
	type CommunityChannelsConnection {
		pageInfo: PageInfo!
		edges: [CommunityChannelEdge!]
	}

	type CommunityChannelEdge {
		node: Channel!
	}

	type CommunityMembersConnection {
		pageInfo: PageInfo!
		edges: [CommunityMemberEdge!]
	}

	type CommunityMemberEdge {
		cursor: String!
		node: User!
	}

	type CommunityThreadsConnection {
		pageInfo: PageInfo!
		edges: [CommunityThreadEdge!]
	}

	type CommunityThreadEdge {
		cursor: String!
		node: Thread!
	}

	type CommunityMetaData {
		members: Int
		channels: Int
	}

	type SlackImport {
		members: String
		teamName: String
		sent: Date
	}

	input CreateCommunityInput {
		name: String!
		slug: String!
		description: String!
		website: String
		file: File
		coverFile: File
	}

	input EditCommunityInput {
		name: String
		description: String
		website: String
		file: File
		coverFile: File
		communityId: ID!
	}

	input EmailInviteContactInput {
		email: String!
		firstName: String
		lastName: String
	}

	input EmailInvitesInput {
		id: ID!
		contacts: [ EmailInviteContactInput ]
		customMessage: String
	}

	input SendSlackInvitesInput {
		id: ID!
		customMessage: String
	}

	type Community {
		id: ID!
		createdAt: Date!
		name: String!
		slug: String!
		description: String!
		website: String
		profilePhoto: String
		coverPhoto: String
		isOwner: Boolean
		isMember: Boolean
		isModerator: Boolean
		isBlocked: Boolean
		pinnedThreadId: String
		communityPermissions: CommunityPermissions!
		channelConnection: CommunityChannelsConnection!
		memberConnection(first: Int = 20, after: String): CommunityMembersConnection!
		threadConnection(first: Int = 10, after: String): CommunityThreadsConnection!
		metaData: CommunityMetaData
		slackImport: SlackImport
		invoices: [Invoice]
	}

	extend type Query {
		community(id: ID, slug: String): Community
		topCommunities(amount: Int = 20): [Community!]
		recentCommunities: [Community!]
		searchCommunities(string: String): [Community]
		searchCommunityThreads(communityId: ID!, searchString: String): [Thread]
	}

	extend type Mutation {
		createCommunity(input: CreateCommunityInput!): Community
		editCommunity(input: EditCommunityInput!): Community
		deleteCommunity(communityId: ID!): Boolean
		toggleCommunityMembership(communityId: ID!): Community
		sendSlackInvites(input: SendSlackInvitesInput!): Community
		sendEmailInvites(input: EmailInvitesInput!): Boolean
		pinThread(threadId: ID!, communityId: ID!, value: String): Community
	}
`;

module.exports = Community;
