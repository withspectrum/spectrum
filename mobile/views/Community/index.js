// @flow
import React, { Component, Fragment } from 'react';
import { Text, View, StatusBar } from 'react-native';
import compose from 'recompose/compose';
import { withNavigation } from 'react-navigation';
import {
  getCommunityById,
  type GetCommunityType,
} from '../../../shared/graphql/queries/community/getCommunity';
import getCommunityThreads from '../../../shared/graphql/queries/community/getCommunityThreadConnection';
import ViewNetworkHandler from '../../components/ViewNetworkHandler';
import ThreadFeed from '../../components/ThreadFeed';
import { ThreadListItem } from '../../components/Lists';
import { getThreadById } from '../../../shared/graphql/queries/thread/getThread';

import {
  Wrapper,
  CoverPhoto,
  CoverPhotoFill,
  CoverPhotoContainer,
  ProfilePhoto,
  ProfilePhotoContainer,
  ProfileDetailsContainer,
  Name,
  Description,
  ThreadFeedDivider,
} from './style';

type Props = {
  isLoading: boolean,
  hasError: boolean,
  navigation: Object,
  data: {
    community?: GetCommunityType,
  },
};

const RemoteThreadItem = compose(getThreadById, withNavigation)(
  ({ data, navigation }) => {
    if (data.loading) return <Text>Loading...</Text>;
    if (!data.thread) return null;
    return (
      <ThreadListItem
        activeCommunity={data.thread.community.id}
        thread={data.thread}
        navigation={navigation}
        onPress={() =>
          navigation.navigate({
            routeName: `Thread`,
            key: data.thread.id,
            params: { id: data.thread.id },
          })
        }
      />
    );
  }
);

const CommunityThreadFeed = compose(getCommunityThreads)(ThreadFeed);

class Community extends Component<Props> {
  render() {
    const { data: { community }, isLoading, hasError } = this.props;

    if (community) {
      return (
        <Wrapper>
          <StatusBar barStyle="light-content" />

          <CommunityThreadFeed
            id={community.id}
            activeCommunity={community.id}
            ListHeaderComponent={
              <Fragment>
                <CoverPhotoContainer>
                  {community.coverPhoto ? (
                    <CoverPhoto
                      resizeMode={'cover'}
                      source={{ uri: community.coverPhoto }}
                    />
                  ) : (
                    <CoverPhotoFill />
                  )}
                </CoverPhotoContainer>

                <ProfilePhotoContainer>
                  <ProfilePhoto source={{ uri: community.profilePhoto }} />
                </ProfilePhotoContainer>

                <ProfileDetailsContainer>
                  <Name>{community.name}</Name>
                  <Description>{community.description}</Description>
                </ProfileDetailsContainer>

                <ThreadFeedDivider />

                {community.pinnedThreadId && (
                  <RemoteThreadItem
                    id={community.pinnedThreadId}
                    activeCommunity={community.id}
                  />
                )}
                {community.watercoolerId && (
                  <RemoteThreadItem
                    id={community.watercoolerId}
                    activeCommunity={community.id}
                  />
                )}
              </Fragment>
            }
          />
        </Wrapper>
      );
    }

    if (isLoading) {
      return (
        <Wrapper>
          <View testID="e2e-community">
            <Text>Loading...</Text>
          </View>
        </Wrapper>
      );
    }

    if (hasError) {
      return (
        <Wrapper>
          <View testID="e2e-community">
            <Text>Error!</Text>
          </View>
        </Wrapper>
      );
    }

    return null;
  }
}

export default compose(getCommunityById, ViewNetworkHandler)(Community);
