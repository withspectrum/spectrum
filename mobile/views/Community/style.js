// @flow
import styled from 'styled-components/native';
import { Dimensions } from 'react-native';

const width = Dimensions.get('window').width;

export const Wrapper = styled.View`
  background-color: ${props => props.theme.bg.default};
  flex: 1;
  align-items: center;
  justify-content: center;
  align-self: stretch;
  width: ${width};
`;

export const CoverPhotoContainer = styled.View`
  width: ${width};
  height: 124px;
  display: flex;
  align-self: stretch;
`;

export const CoverPhoto = styled.Image`
  width: ${width};
  height: 124px;
`;

export const CoverPhotoFill = styled.View`
  width: ${width};
  height: 124px;
  background: ${props => props.theme.brand.alt};
`;

export const ProfilePhotoContainer = styled.View`
  width: ${width};
  display: flex;
  justify-content: flex-start;
  padding: 0 16px;
  position: relative;
  top: -32px;
  margin-bottom: -24px;
`;

export const ProfilePhoto = styled.Image`
  width: 72px;
  height: 72px;
  border-radius: 12px;
  border-width: 3px;
  border-color: ${props => props.theme.bg.default};
  shadow-color: ${props => props.theme.bg.default};
  shadow-opacity: 1;
  shadow-radius: 4;
`;

export const ProfileDetailsContainer = styled.View`
  display: flex;
  flex-direction: column;
  width: ${width};
  padding: 0 24px;
`;

export const Name = styled.Text`
  font-weight: 800;
  color: ${props => props.theme.text.default};
  font-size: 20px;
`;

export const Description = styled.Text`
  font-weight: 400;
  color: ${props => props.theme.text.default};
  font-size: 16px;
  margin-top: 8px;
`;

export const ThreadFeedDivider = styled.View`
  height: 32px;
  border-bottom-color: ${props => props.theme.bg.border};
  border-bottom-width: 1px;
`;
