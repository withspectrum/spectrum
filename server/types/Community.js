const Community = /* GraphQL */ `
	type CommunityFrequenciesConnection {
		pageInfo: PageInfo!
		edges: [CommunityFrequencyEdge!]
	}

	type CommunityFrequencyEdge {
		node: Frequency!
	}

	type CommunityMembersConnection {
		pageInfo: PageInfo!
		edges: [CommunityMemberEdge!]
	}

	type CommunityMemberEdge {
		cursor: String!
		node: User!
	}

	type CommunityStoriesConnection {
		pageInfo: PageInfo!
		edges: [CommunityStoryEdge!]
	}

	type CommunityStoryEdge {
		cursor: String!
		node: Story!
	}

	type CommunityMetaData {
		members: Int
		frequencies: Int
	}

	type Community {
		id: ID!
		createdAt: Date!
		name: String!
		slug: String!
		frequencyConnection: CommunityFrequenciesConnection!
		memberConnection(first: Int = 10, after: String): CommunityMembersConnection!
		storyConnection(first: Int = 10, after: String): CommunityStoriesConnection!
		metaData: CommunityMetaData
	}

	extend type Query {
		community(id: ID, slug: String): Community
	}

	extend type Mutation {
		createCommunity(name: String!, slug: String!): Community
	}
`;

module.exports = Community;
