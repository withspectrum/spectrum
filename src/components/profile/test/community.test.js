// @flow
import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Community from '../community';

const data = {
  title: 'Test community',
  subtitle: 'Whats up',
  description: 'Testing the community profile',
  id: 'asdf-123',
};

it('should render with no size', () => {
  const wrapper = shallow(<Community data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should render with a mini size', () => {
  const wrapper = shallow(<Community size="mini" data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should render with a small size', () => {
  const wrapper = shallow(<Community size="small" data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should render with a medium size', () => {
  const wrapper = shallow(<Community size="medium" data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should render with a large size', () => {
  const wrapper = shallow(<Community size="large" data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should render with a full size', () => {
  const wrapper = shallow(<Community size="full" data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});
