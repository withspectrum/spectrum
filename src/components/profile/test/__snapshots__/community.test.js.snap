exports[`test should render with a full size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the community profile",
      "id": "asdf-123",
      "subtitle": "Whats up",
      "title": "Test community",
    }
  }
  size="full">
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test community
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
  <styled.p>
    Testing the community profile
  </styled.p>
  <Styled()>
    <Styled()>
      Follow
    </Styled()>
  </Styled()>
  <Apollo(branch(pure(MetaDataPure)))
    id="asdf-123"
    type="community" />
</pure(CardPure)>
`;

exports[`test should render with a large size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the community profile",
      "id": "asdf-123",
      "subtitle": "Whats up",
      "title": "Test community",
    }
  }
  size="large">
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test community
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
  <styled.p>
    Testing the community profile
  </styled.p>
  <Styled()>
    <Styled()>
      Follow
    </Styled()>
  </Styled()>
  <Apollo(branch(pure(MetaDataPure)))
    id="asdf-123"
    type="community" />
</pure(CardPure)>
`;

exports[`test should render with a medium size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the community profile",
      "id": "asdf-123",
      "subtitle": "Whats up",
      "title": "Test community",
    }
  }
  size="medium">
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test community
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
  <styled.p>
    Testing the community profile
  </styled.p>
  <Styled()>
    <Styled()>
      Follow
    </Styled()>
  </Styled()>
</pure(CardPure)>
`;

exports[`test should render with a mini size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the community profile",
      "id": "asdf-123",
      "subtitle": "Whats up",
      "title": "Test community",
    }
  }
  size="mini">
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test community
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
</pure(CardPure)>
`;

exports[`test should render with a small size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the community profile",
      "id": "asdf-123",
      "subtitle": "Whats up",
      "title": "Test community",
    }
  }
  size="small">
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test community
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
  <Styled()>
    <Styled()>
      Follow
    </Styled()>
  </Styled()>
</pure(CardPure)>
`;

exports[`test should render with no size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the community profile",
      "id": "asdf-123",
      "subtitle": "Whats up",
      "title": "Test community",
    }
  }>
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test community
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
</pure(CardPure)>
`;
